pkg/providers/tinkerbell/upgrade.go

```diff
@@ -244,7 +245,22 @@ func (p *Provider) ValidateNewSpec(ctx context.Context, cluster *types.Cluster,
 	oSpec := prevDatacenterConfig.Spec
 	nSpec := p.datacenterConfig.Spec
 
-	prevMachineConfigRefs := machineRefSliceToMap(prevSpec.MachineConfigRefs())
+	currCPMachineRef := prevSpec.Spec.ControlPlaneConfiguration.MachineGroupRef
+	desiCPMachineRef := clusterSpec.Cluster.Spec.ControlPlaneConfiguration.MachineGroupRef
+	if !currCPMachineRef.Equal(desiCPMachineRef) {
+		return errors.New("control plane machine config reference is immutable")
+	}
+
+	err = validateWorkerNodeGroupMachineConfigRefsUnchanged(prevSpec.Spec.WorkerNodeGroupConfigurations,
```

Comment:
As it's not clear to me, I would be interested in a code comment or similar to describe why we might want these to be unchanged.

Reply:
If I knew concretely, I would love to add it, sadly I don't. This PR merely adds a missing piece of the logic to correctly enforce the constraint.

Reply:
If I knew concretely, I would love to add it, sadly I don't. This PR merely adds a missing piece of the logic to correctly enforce the constraint.